### PR TITLE
Cosmos Spark: Collision resolution on schema inference

### LINKED ISSUE #21818: OLTP Connector 3 is unable to infer schema correctly schema and thus failing with number-format-exception for particular cosmos-collection
CODE : 
val cosmosEndpoint = <>
val cosmosMasterKey = <>
val cosmosDatabaseName = <>
val cosmosContainerName2 = <particular-collection>

val cfgWithAutoSchemaInference2 = Map("spark.cosmos.accountEndpoint" -> cosmosEndpoint,
  "spark.cosmos.accountKey" -> cosmosMasterKey,
  "spark.cosmos.database" -> cosmosDatabaseName,
  "spark.cosmos.container" -> cosmosContainerName2,
  "spark.cosmos.read.inferSchema.enabled" -> "true"
)

var df2 = spark.read.format("cosmos.oltp").options(cfgWithAutoSchemaInference2).load()

df2.printSchema()

RESULT Provided by cosmos connector 3 : 
root
 |-- dataDetails: string (nullable = true)
 |-- sensorDetails: string (nullable = false)
 |-- partitionKey: string (nullable = false)
 |-- id: string (nullable = false)
 |-- msgSchema: struct (nullable = false)
 |    |-- msgVersion: string (nullable = false)

Expected Result (tested out from cosmos connector 2) : 
root
 |-- _attachments: string (nullable = true)
 |-- _etag: string (nullable = true)
 |-- _rid: string (nullable = true)
 |-- _self: string (nullable = true)
 |-- _ts: integer (nullable = true)
 |-- id: string (nullable = true)
 |-- msgSchema: struct (nullable = true)
 |    |-- msgVersion: string (nullable = true)
 |-- dataDetails: struct (nullable = true)
 |    |-- rackName: string (nullable = true)
 |    |-- caseName: string (nullable = true)
 |    |-- circuitName: string (nullable = true)
 |    |-- sensorDescription: string (nullable = true)
 |-- partitionKey: string (nullable = true)
 |-- sensorData: struct (nullable = true)
 |    |-- networkType: string (nullable = true)
 |    |-- sensorStatus: struct (nullable = true)
 |    |    |-- mappedReading: null (nullable = true)
 |    |    |-- reading: string (nullable = true)
 |    |    |-- units: string (nullable = true)
 |    |    |-- unResolvedReading: string (nullable = true)
 |    |-- rackIndex: integer (nullable = true)
 |    |-- vendorType: string (nullable = true)
 

Query - display(df)
RESULT :  
NumberFormatException: For input string: "3612611867"

NOTE : CONNECTOR 2 is working fine for same cosmos collection (JAVA SDK older version is working fine but SDK 4 is also throwing same exception)

### PR DESCRIPTION
**Pull request by ealsur**
When schema inference goes through documents that have the same property but different schema, for example:

```json
{
    "property": {
        "nested1": "some Value"
    }
}
```

and 

```json
{
    "property": {
        "nested2": "some other Value"
    }
}
```

It's currently infering `StringType` because they are different definitions.

This PR incorporates the type coercion resolution already present in Spark Connector 2, which would produce a schema where the property `property` has a `StructType` with 2 fields.

Closes #21818

### CODE REVIEW

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala [hunk 1]
```diff
@@ -175,4 +178,48 @@ private object CosmosTableSchemaInferrer
...
+  private def compatibleType(t1: DataType, t2: DataType): DataType = {
+    //TypeCoercion.findTightestCommonTypeOfTwo(t1, t2) match {
+    TypeCoercion.findTightestCommonType(t1, t2) match {
+      case Some(commonType) => commonType
+
+      case None =>
+        // t1 or t2 is a StructType, ArrayType, or an unexpected type.
+        (t1, t2) match {
+          case (other: DataType, NullType) => other
+          case (NullType, other: DataType) => other
```
Comment (moderakh) [inline sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala:203]: I wonder if we can add some test cases for all branches here including null type

**Review by moderakh — APPROVED** [review]
LGTM.

could you please consider adding unit tests for all branches in type collision resolution?

File: sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala (hunk 1 above)
Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala:203]: The test was already there before `it should "map duplicate properties when one is nullable"`

Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala:203]: I will add to it on the other direction (first node having null, second having a value) which should take care of the other case

Comment (ealsur) [reply sdk/cosmos/azure-cosmos-spark_3-1_2-12/src/main/scala/com/azure/cosmos/spark/CosmosTableSchemaInferrer.scala:203]: Added another test for it.

**Comment by tyagi420** [comment]
Thanks you guys, I assume fixes will be available in 4.2.0-beta.1 ?
